source/event_stream.c

```diff
@@ -0,0 +1,1118 @@
+                  size_t *processed) {
+
+    size_t current_pos = decoder->message_pos;
+
+    size_t length_read = current_pos - decoder->current_header_value_offset;
+
+    if (!length_read) {
+        /* save an allocation, this can only happen if the data we were handed is larger than the length of the header value. we don't really need to handle offsets in this case.
+        This expects the user is living by the contract that they cannot act like they own this memory beyond the lifetime of their callback, and they should not mutate it */
+        if (len >= decoder->current_header.header_value_len) {
+            /* this part works regardless of type since the layout of the union will line up. */
+            decoder->current_header.header_value.variable_len_val = data;
+            decoder->current_header.value_owned = 0;
+            decoder->on_header(decoder, &decoder->prelude, &decoder->current_header, decoder->user_context);
+            *processed += decoder->current_header.header_value_len;
+            decoder->message_pos += decoder->current_header.header_value_len;
+            decoder->running_crc = aws_checksums_crc32(data, (int) decoder->current_header.header_value_len,
+                                                       decoder->running_crc);
+
+            reset_header_state(decoder, 1);
+            decoder->state = headers_state;
+            return AWS_OP_SUCCESS;
+        }
+
+        /* a possible optimization later would be to only allocate this once, and then keep reusing the same buffer. for subsequent messages.*/
+        if (decoder->current_header.header_value_type == AWS_EVENT_STREAM_HEADER_BYTE_BUF ||
+            decoder->current_header.header_value_type == AWS_EVENT_STREAM_HEADER_STRING) {
+            decoder->current_header.header_value.variable_len_val = aws_mem_acquire(decoder->alloc,
+                                                                                    decoder->current_header.header_value_len);
+
+            if (!decoder->current_header.header_value.variable_len_val) {
+                return aws_raise_error(AWS_ERROR_OOM);
+            }
+
+            decoder->current_header.value_owned = 1;
+        }
+    }
+
+    size_t max_read =
+            len >= decoder->current_header.header_value_len - length_read ? decoder->current_header.header_value_len -
+                                                                            length_read : len;
+    const uint8_t *header_value_alias = decoder->current_header.header_value_type == AWS_EVENT_STREAM_HEADER_BYTE_BUF ||
+                                        decoder->current_header.header_value_type == AWS_EVENT_STREAM_HEADER_STRING ?
+                                        decoder->current_header.header_value.variable_len_val
+                                                                                                                    : decoder->current_header.header_value.static_val;
```

Comment:
this function needs some work. It's not very readable. decoder->current_header is repeated 500 times. the formatting is off in multiple places.